pallets/roles/src/lib.rs

```diff
@@ -0,0 +1,239 @@
+		/// Slashed validator.
+		Slashed { account: T::AccountId, amount: BalanceOf<T> },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Role has already been assigned to provided validator.
+		RoleAlreadyAssigned,
+		/// No role assigned to provided validator.
+		RoleNotAssigned,
+		/// Insufficient bond to become a validator.
+		InsufficientBond,
+		/// Stash controller account already added to Ledger
+		AlreadyPaired,
+		/// Stash controller account not found in ledger
+		InvalidStashController,
+	}
+
+	/// Map from all "controller" accounts to the info regarding the staking.
+	#[pallet::storage]
+	#[pallet::getter(fn ledger)]
+	pub type Ledger<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AccountId, RoleStakingLedger<T>>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn account_role)]
+	/// Mapping of resource to bridge index
+	pub type AccountRolesMapping<T: Config> = StorageMap<_, Blake2_256, T::AccountId, RoleType>;
+
+	/// The minimum active bond to become and maintain the role.
+	#[pallet::storage]
+	#[pallet::getter(fn min_active_bond)]
+	pub(super) type MinActiveBond<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight({0})]
+		#[pallet::call_index(0)]
+		pub fn assign_role(
+			origin: OriginFor<T>,
+			#[pallet::compact] bond_value: BalanceOf<T>,
+			role: RoleType,
+		) -> DispatchResult {
+			let stash_account = ensure_signed(origin)?;
+			// check if role is already assigned.
+			ensure!(
+				!AccountRolesMapping::<T>::contains_key(&stash_account),
+				Error::<T>::RoleAlreadyAssigned
+			);
+			// check if stash account is already paired.
+			if <Ledger<T>>::contains_key(&stash_account) {
+				return Err(Error::<T>::AlreadyPaired.into())
+			}
+			// check if min active bond is met.
+			let min_active_bond = MinActiveBond::<T>::get();
+			if bond_value < min_active_bond.into() {
+				return Err(Error::<T>::InsufficientBond.into())
+			}
+			// Bond with stash account.
+			let stash_balance = T::Currency::free_balance(&stash_account);
+			let value = bond_value.min(stash_balance);
+
+			// update ledger.
+			let item = RoleStakingLedger { stash: stash_account.clone(), total_locked: value };
+			Self::update_ledger(&stash_account, &item);
+
+			Self::deposit_event(Event::<T>::Bonded {
+				account: stash_account.clone(),
+				amount: value,
+			});
+
+			// Add role mapping for the stash account.
+			AccountRolesMapping::<T>::insert(&stash_account, role);
+			Self::deposit_event(Event::<T>::RoleAssigned { account: stash_account.clone(), role });
+			Ok(())
+		}
+
+		#[pallet::weight({0})]
+		#[pallet::call_index(1)]
+		pub fn clear_role(origin: OriginFor<T>, role: RoleType) -> DispatchResult {
+			let stash_account = ensure_signed(origin)?;
+			// check if role is assigned.
+			ensure!(
+				Self::validate_role(stash_account.clone(), role.clone()),
+				Error::<T>::RoleNotAssigned
+			);
+			// TODO: Call jobs manager to remove the services.
```

Comment:
Let's create a task for TODOs as well and link the issue in code so we know when it has been addressed and can add details. These types of TODOs are of maximal importance.